### ISSUE TITLE
20 - Promise.all - asking for a weird behavior

### ISSUE THREAD

**Issue by Drincann** [type-challenges/type-challenges#20852]
The following code will fail in the case 3:
```ts
declare function PromiseAll<A extends readonly unknown[]>(values: A): Promise<{
  -readonly [key in keyof A]: Awaited<A[key]>
}>

// test
const promiseAllTest3 = PromiseAll([1, 2, Promise.resolve(3)]) // Promise<number[]>
```

But when I change the generic `readonly unknown[]` to `(readonly unknown[]) | []` like below，everything will be ok.

```ts
declare function PromiseAll<A extends (readonly unknown[]) | [] /* change here */>(values: A): Promise<{
  -readonly [key in keyof A]: Awaited<A[key]>
}>

// test
const promiseAllTest3 = PromiseAll([1, 2, Promise.resolve(3)]) // Promise<[number, number, number]>
```

I don't understand why adding a constraint of '[]' to generic 'A' will affect the case 3. I can't associate them at all.

In addition, the correct implementation is the standard implementation of lib.es2015.promise.d.ts, which I found in vscode

```ts
// lib.es2015.promise.d.ts
/**
 * Creates a Promise that is resolved with an array of results when all of the provided Promises
 * resolve, or rejected when any Promise is rejected.
 * @param values An array of Promises.
 * @returns A new Promise.
 */
all<T extends readonly unknown[] | []>(values: T): Promise<{ -readonly [P in keyof T]: Awaited<T[P]> }>;
```

**Comment by Drincann** [comment]
solved here [stackoverflow#74848194](https://stackoverflow.com/questions/74848194/how-does-the-type-defination-of-promise-all-work-well-in-this-case)

**Comment by L1atte** [comment]
I am sorry to trouble you and i want to know what the code `-readonly` means here

> `all<T extends readonly unknown[] | []>(values: T): Promise<{ -readonly [P in keyof T]: Awaited<T[P]> }>;`

**Comment by Drincann** [comment]
Yes, in the mapping type, you can use it before the modifier to remove (if any) the modifier.
Refer to [Mapping Modifiers](https://www.typescriptlang.org/docs/handbook/2/mapped-types.html#mapping-modifiers).

**Comment by L1atte** [comment]
> Yes, in the mapping type, you can use it before the modifier to remove (if any) the modifier. Refer to [Mapping Modifiers](https://www.typescriptlang.org/docs/handbook/2/mapped-types.html#mapping-modifiers).

Thanks for your answer! I have never used it before :(

**Comment by wkz2003** [comment]
兄弟我看了你贴出的stackoverflow的回答，可能我英语不是很好还不是很理解。请教一下两个个问题：
1.回答里写那个函数参数里面的readonly是一种转换，不是约束，是这个意思对吧？
2.我还是不太明白为什么会推断出两种写法tuple类型。（上下文推断怎么理解）

**Comment by wkz2003** [comment]
> 

那extends readonly unknown[]这个不是也算元组吗，还是说readonly不一定是元组呢

**Comment by danvk** [comment]
With [TypeScript 5.0's `const` type parameters][1], you can do better:

```ts
declare function PromiseAll<const A extends readonly unknown[]>(values: A): Promise<{
  -readonly [key in keyof A]: Awaited<A[key]>
}>

// test
const promiseAllTest3 = PromiseAll([1, 2, Promise.resolve(3)]) // Promise<number[]>
//    ^? const promiseAllTest3: Promise<[1, 2, number]>
```

See #28594

[1]: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-0.html#const-type-parameters
